engine/src/main/java/org/terasology/logic/players/LocalPlayerSystem.java

```diff
@@ -160,6 +170,43 @@ private void processInput(EntityRef entity, CharacterMovementComponent character
         jump = false;
     }
 
+    // Reduces height and eyeHeight by crouch_fraction and changes MovementMode.
+    // TODO: Add animation while crouching
+    // TODO: Shift crouch_fraction = 0.5f outside
+    private void crouchPlayer(EntityRef entity) {
+        ClientComponent clientComp = entity.getComponent(ClientComponent.class);
+        GazeMountPointComponent gazeMountPointComponent = clientComp.character.getComponent(GazeMountPointComponent.class);
+        float height = clientComp.character.getComponent(CharacterMovementComponent.class).height;
+        float eyeHeight = gazeMountPointComponent.translate.getY();
+        movementDebugCommands.playerHeight(localPlayer.getClientEntity(), height * 0.5f);
+        movementDebugCommands.playerEyeHeight(localPlayer.getClientEntity(), eyeHeight * 0.5f);
+        clientComp.character.send(new SetMovementModeEvent(MovementMode.CROUCHING));
+    }
+
+    // Checks if there is an impenetrable block above,
+    // Raises a Notification "Cannot stand here!" if present
+    // If not present, increases height and eyeHeight by crouch_fraction and changes MovementMode.
+    // TODO: Add animation while standing
+    private void standPlayer(EntityRef entity) {
+        ClientComponent clientComp = entity.getComponent(ClientComponent.class);
+        GazeMountPointComponent gazeMountPointComponent = clientComp.character.getComponent(GazeMountPointComponent.class);
+        float height = clientComp.character.getComponent(CharacterMovementComponent.class).height;
+        float eyeHeight = gazeMountPointComponent.translate.getY();
+        Vector3f pos = entity.getComponent(LocationComponent.class).getWorldPosition();
+        // Check for collision when rising
+        CharacterCollider collider = physics.getCharacterCollider(clientComp.character);
+        // height used below = standing_height - crouch_fraction * standing_height
+        Vector3f to = new Vector3f(pos.x, pos.y + height, pos.z);
+        SweepCallback callback = collider.sweep(pos, to, VERTICAL_PENETRATION_LEEWAY, -1f);
+        if (callback.hasHit()) {
+            entity.send(new NotificationMessageEvent("Cannot stand here!", entity));
```

Comment:
Tiny clarification: "Cannot stand *up* here" ? Seems to me like it would slightly better Indicate an inability to take an action to stand up there.